File: code/backend/python/app/core/services/story/story_catalog_export_service.py
from typing import List
from app.core.models.story.catalog_export_result import CatalogExportResult, CatalogFile
from app.core.ports.story.catalog_writer_port import CatalogWriterPort
from app.core.ports.story.story_catalog_export_port import (
    StoryCatalogExportPort, CatalogNotConfiguredError, catalog_path_for)
from app.core.ports.story.story_query_port import StoryQueryPort


class StoryCatalogExportService(StoryCatalogExportPort):
    """v0.37.6 — writes the public story list of every configured language through the writer."""

    def __init__(self, query_port: StoryQueryPort, writer: CatalogWriterPort, langs: List[str] | None = None):
        self.query_port = query_port
        self.writer = writer
        self.langs = list(langs) if langs else ["en"]

    def export_catalog(self) -> CatalogExportResult:
        if not self.writer.is_configured():
            raise CatalogNotConfiguredError()
        files = []
        for lang in self.langs:
            stories = self.query_port.list_public_stories(lang)
            path = catalog_path_for(lang)
            size = self.writer.write(path, stories)
            files.append(CatalogFile(lang=lang, path=path, count=len(stories), bytes=size))
        return CatalogExportResult(target=self.writer.target(), files=files)

File: code/backend/java/core/src/main/java/games/paths/core/service/story/StoryCatalogExportService.java
package games.paths.core.service.story;

import java.util.ArrayList;
import java.util.List;

import games.paths.core.model.story.CatalogExportResult;
import games.paths.core.model.story.StorySummary;
import games.paths.core.port.story.CatalogWriterPort;
import games.paths.core.port.story.StoryCatalogExportPort;
import games.paths.core.port.story.StoryQueryPort;

/**
 * StoryCatalogExportService - v0.37.6 writes the public story list of every
 * configured language through the {@link CatalogWriterPort}.
 */
public class StoryCatalogExportService implements StoryCatalogExportPort {

    private final StoryQueryPort storyQueryPort;
    private final CatalogWriterPort writer;
    private final List<String> langs;

    public StoryCatalogExportService(StoryQueryPort storyQueryPort, CatalogWriterPort writer, List<String> langs) {
        this.storyQueryPort = storyQueryPort;
        this.writer = writer;
        this.langs = langs == null || langs.isEmpty() ? List.of("en") : List.copyOf(langs);
    }

    @Override
    public CatalogExportResult exportCatalog() {
        if (!writer.isConfigured()) {
            throw new CatalogNotConfiguredException();
        }
        List<CatalogExportResult.CatalogFile> files = new ArrayList<>();
        for (String lang : langs) {
            List<StorySummary> stories = storyQueryPort.listPublicStories(lang);
            String path = StoryCatalogExportPort.pathFor(lang);
            int bytes = writer.write(path, stories);
            files.add(new CatalogExportResult.CatalogFile(lang, path, stories.size(), bytes));
        }
        return new CatalogExportResult(writer.target(), files);
    }
}
